agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
+/// It cannot be assumed that this will successfully update Instance on first try since Device Plugins on other nodes may be simultaneously trying to update the Instance.
+/// This returns an error if slot does not need to be updated or `MAX_INSTANCE_UPDATE_TRIES` attempted.
+async fn try_update_instance_device_usage(
+    device_usage_id: &str,
+    node_name: &str,
+    instance_name: &str,
+    instance_namespace: &str,
+    device_usage_kind: DeviceUsageKind,
+    kube_interface: Arc<impl KubeInterface>,
+) -> Result<(), Status> {
+    let mut instance: InstanceSpec;
+    for x in 0..MAX_INSTANCE_UPDATE_TRIES {
+        // Grab latest instance
+        match kube_interface
+            .find_instance(instance_name, instance_namespace)
+            .await
+        {
+            Ok(instance_object) => instance = instance_object.spec,
+            Err(_) => {
+                trace!(
+                    "try_update_instance_device_usage - could not find Instance {}",
+                    instance_name
+                );
+                return Err(Status::new(
+                    Code::Unknown,
                     format!("Could not find Instance {}", instance_name),
                 ));
             }
         }
 
         // Update the instance to reserve this slot for this node iff it is available and not already reserved for this node.
-        if slot_available_to_reserve(device_usage_id, node_name, &instance)? {
-            instance
-                .device_usage
-                .insert(device_usage_id.to_string(), node_name.to_string());
+        let device_usage_string = instance.device_usage.get(device_usage_id);
+        if device_usage_string.is_none() {
+            // No corresponding id found
+            trace!(
+                "try_update_instance_device_usage - could not find {} id in device_usage",
+                device_usage_id
+            );
+            return Err(Status::new(
+                Code::Unknown,
+                "Could not find device usage slot",
+            ));
+        }
 
-            if let Err(e) = kube_interface
-                .update_instance(&instance, instance_name, instance_namespace)
-                .await
-            {
-                if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
-                    trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
-                    return Err(Status::new(Code::Unknown, "Could not update Instance"));
+        let device_usage = DeviceUsage::from_str(device_usage_string.unwrap()).map_err(|_| {
+            Status::new(
+                Code::Unknown,
+                format!(
+                    "Fails to parse {} to DeviceUsage ",
+                    device_usage_string.unwrap()
+                ),
+            )
+        })?;
+        match get_device_usage_state(&device_usage, node_name) {
```

Comment:
might be nice to pull this `match ...` into a function and add some tests to validate what happens when Free is encountered vs when Configuration or Instance (for this node and for other nodes) vs when ReservedForOthers.

Reply:
we have test cases to cover that